### ISSUE TITLE
[REVIEW]: fib-tf: A TensorFlow-based Cardiac Electrophysiology Simulator

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#719]
**Submitting author:** @siravan (<a href="http://orcid.org/0000-0003-2132-1543">Shahriar Iravanian</a>)
**Repository:** <a href="https://github.com/siravan/fib_tf" target ="_blank">https://github.com/siravan/fib_tf</a>
**Version:** v1.0.0
**Editor:** @Kevin-Mattheus-Moerman
**Reviewer:** @mgalloy, @leios
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.1256950" target="_blank">10.5281/zenodo.1256950</a>

## Status

[![status](http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222/status.svg)](http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222"><img src="http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222/status.svg)](http://joss.theoj.org/papers/6a7581d6253014e4a091947cc0a30222)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/siravan/fib_tf" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mgalloy & @leios, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/siravan/fib_tf">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0.0)?
- [x] **Authorship:** Has the submitting author (@siravan) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/siravan/fib_tf">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0.0)?
- [x] **Authorship:** Has the submitting author (@siravan) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @mgalloy, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00719/joss.00719/10.21105.joss.00719.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@siravan this is where the review will take place. @mgalloy, @leios thanks for acting as reviewers for this submission. There are tickboxes at the top here to guide you through the review process. 

Let me know if I can help. 
:rocket: :robot:

**Comment by mgalloy** [comment]
I believe the DOI for the "Reconstruction of the action potential of ventricular myocardial fibres" article in your References is "https://doi.org/10.1113/jphysiol.1977.sp011853". I don't see a DOI for the TensorFlow reference.

**Comment by siravan** [comment]
Thanks for the comments. The shebangs are fixed. I also added the doi you mentioned. The TensorFlow whitepaper does not seem to have a doi, as it is posted on the TensorFlow website.

**Comment by mgalloy** [comment]
@siravan A couple questions:

- Are you going to make a release via GitHub? Currently, there are no releases.
- Do you have any guidelines for contributing? asking for support? If not, a short section in your README could probably satisfy the "Community guidelines" above.

**Comment by siravan** [comment]
Yes, I plan to make a release by the end of the review process, when things are more stable.

Thanks for pointing to the lack of contributing guidelines in the README. I will add one soon. 

I welcome contributions from a diverse background, as the man goal of fib-tf is experimentation with different ideas regarding solving ODEs with TensorFlow.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #719 with the following error: 

   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    13    0    13    0     0     90      0 --:--:-- --:--:-- --:--:--    90
Error reading bibliography ./paper.bib (line 68, column 1):
unexpected "i"
expecting space, ",", white space or "}"
Error running filter pandoc-citeproc:
Filter returned error status 1
Looks like we failed to compile the PDF

**Comment by Kevin-Mattheus-Moerman** [comment]
@siravan perhaps when you pasted in the DOI you deleted a comma? Not sure if this is it but merge my pull request and we can try `@whedon generate pdf` again, btw you can ask whedon to do this too.

**Comment by siravan** [comment]
Yes, I missed a comma. It is fixed now.

**Comment by Kevin-Mattheus-Moerman** [comment]
@siravan, in terms of community guidelines and contributing. I would recommend adding a `CONTRIBUTING.md` file and also a `CODE_OF_CONDUCT.md` file. 
You can base the latter on the Contributor Covenant, e.g. version 1.4, available at http://contributor-covenant.org/version/1/4

For more information about community guidelines: https://help.github.com/articles/github-community-guidelines/

Also this link offers example CONTRIBUTING.md files from other projects:
https://help.github.com/articles/setting-guidelines-for-repository-contributors/

Let me know if you have questions.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00719/joss.00719/10.21105.joss.00719.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@siravan I see you already added a CONTRIBUTING.md file. I recommend adding a code of conduct file as well.

**Comment by Kevin-Mattheus-Moerman** [comment]
@mgalloy, @leios can you summarize where we stand with the review process? Thanks

**Comment by siravan** [comment]
A CODE_OF_CONDUCT.md file is added to the master branch.

**Comment by mgalloy** [comment]
@Kevin-Mattheus-Moerman  OK, assuming@siravan makes a GitHub release labeled 1.0, then the only thing I haven't found yet is "Functionality documentation", i.e., API documentation. Am I missing it somewhere @siravan?

**Comment by leios** [comment]
@Kevin-Mattheus-Moerman I apologize for taking so long with this review. I ran into a rather annoying build error when attempting to build tensorflow with JIT and CUDA (https://github.com/tensorflow/tensorflow/issues/10220). It took a while for me to find the solution. I should have been quicker to realize that I needed to use another version of `gcc`.

@mgalloy The documentation can be found in the `docs/` directory under `details.html.` It's pretty good. There is also a small link under the Documentation heading in the README. It's not exactly obvious to find.

Ultimately, the software works as described; however (@siravan), I have a few lingering questions that have not been addressed here:

1. The statement of need wasn't quite clear for me. What are the future applications of this software and where has it already been used in an academic (or non-academic) setting? There was a mention of the systems it can be used for, but no mention of what users would be interested in using this software in-particular or any publications that may result from this software. Is it meant to be used by researchers studying stiff ODE systems or is it a test-case to show that tensorflow can provide decently optimized solutions to such systems?

2. There is a mention of "hand-optimized CUDA code" a lot, but I don't see the code anywhere? I am not sure what you are benchmarking against to get "within a factor of 2-3 of..."

3. So far as I can tell, this software uses the forward Euler method, which is known to be unstable and carry a high degree of error for stiff ODE's, which is the explicit purpose of this work. Because of this, the forward Euler method is not typically used in too many physics applications. Instead, most researchers prefer using Runge-Kutta, Crank-Nicholson, or pseudo-spectral methods. I understand that this application is primarily for electrophysiological systems, but it would seem like a good idea to include other methods to solve the ODE systems to be safe. If it is commonplace to use more error-prone methods for these types of simulations, then maybe it is fine not to address this point; however, I know that there are researchers who will not use anything with an error of less than dt^3.

4. Some of the links in the documentation do not work. Those should probably be fixed.

Please let me know if I missed anything!

**Comment by siravan** [comment]
@leios Thanks for the comments. I appreciate your comments and persistence in installing the JIT enabled TensorFlow. 

1. The initial reason to develop fib_tf was as a test bed for new ideas on cardiac electrophysiology simulation, like adding a new ionic current to a cardiac model. This is actually how we use it in our academic work (no publication yet, but we are working on a paper on enhancing the Courtemanche atrial model). It is much easier to test ideas in a scripting language like python with all its scientific computing ecosystem than in C++. However, fit_tf can also be useful in assessing the feasibility of writing ODE solvers in TensorFlow. 

The main power of TensorFlow is its ability to run on multiple GPUs or even on a distributed system with multiple CPUs, each with one or more GPUs. Writing a C++ code to do this is of course extremely difficult. In this situation, the python code could potentially be faster than the single GPU C++.

2. The "hand-optimized CUDA code" mentioned here is the production code that we use for 3D cardiac modeling (for [example](https://aip.scitation.org/doi/10.1063/1.5000350)). My plan is to refactor the C++ code and CUDA kernles and add the relevant parts to fib_tf. The C++ code base in its current form is too complex and has lots of functionality unrelated to fib_tf.

3. Electrophysiology applications are rather unusual in this regard and are commonly solved using the explicit Euler model (or by implicit solvers on occasion). The reason is the extreme stiffness of models. Higher order explicit solvers, such as Runge-Kutta or Linear Multi-Step methods, are not of much help and one still needs a very small time-step for stability. However, the Rush_larsen technique that we have used here is very helpful and allows us to integrate more than half the state variables (the gating variables) by a closed-form integration formula.

Also please note that TensorFlow has an ODE solver (tf.contrib.integrate.odeint), which implements 5th order Runge-Kutta with adaptive step size control and dense output, using the Dormand-Prince method. If it was logical to use a higher order method, we could have used this functionality.

It should be also noted that TensorFlow has an automatic differentiation capability (needed for the machine learning applications). It is possible to implement high order Taylor series methods by using the automatic differentiation (TensorFlow only has backward differentiation, but there is a trick when one can combine two backward differentiations to obtain a forward one needed in the Taylor series). Again this is not a useful technique for cardiac electrophysiology applications, but is potentially applicable to non-stiff models. 

4. The links are fixed.

**Comment by leios** [comment]
@siravan Thanks for detailed response! I believe that answers all my questions for now, but will take one final look at the code soon.

**Comment by Kevin-Mattheus-Moerman** [comment]
@mgalloy @leios thanks for your review!! :rocket: can you report on any remaining issues or advise on whether you deem this submission acceptable? 
@leios the _Functionality documentation_ box is unticked. Can you summarize what you feel is missing?
Thanks

**Comment by leios** [comment]
@Kevin-Mattheus-Moerman All my boxes are ticked and I think this submission is acceptable. 

@mgalloy is still missing the *functionality documentation* box, though. I think they had trouble finding the documentation, so I tried to show them where it was. I don't know if they have looked at this thread since then.

**Comment by mgalloy** [comment]
The documentation pointed out by @leios is a great explanation . I have checked off the *functionality documentation* and agree this submission is acceptable.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00719/joss.00719/10.21105.joss.00719.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@siravan looks like we are nearly there. I suggest a minor change to the first sentence in your paper. Instead of `on the top of the` consider saying `on top of the`. If you agree with this make the change and render a new pdf here using `@whedon generate pdf`. 
Once you are done with that step please archive a copy of the **reviewed software** using a service like [Zenodo](http://zenodo.org/) ([see here to automate DOI creation on new releases](https://guides.github.com/activities/citable-code/)), or [figshare](https://figshare.com/), and list the DOI of the archived version here.

**Comment by Kevin-Mattheus-Moerman** [comment]
@leios @mgalloy thanks for the excellent and smooth review process! :tada: :rocket: :robot:

**Comment by siravan** [comment]
Thanks a lot for all the help and comments. The paper.md is edited and the repository is archived by Zenodo. DOI is
 [![DOI](https://zenodo.org/badge/111343472.svg)](https://zenodo.org/badge/latestdoi/111343472)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00719/joss.00719/10.21105.joss.00719.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
Great. Thanks @siravan. @arfon we are all set to accept here.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.1256950 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.1256950" target="_blank">10.5281/zenodo.1256950</a> is the archive.

**Comment by arfon** [comment]
@mgalloy, @leios - many thanks for your reviews here and to @Kevin-Mattheus-Moerman for editing this submission ✨ 

@siravan - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00719 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippet:

```
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00719/status.svg)](https://doi.org/10.21105/joss.00719)
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00719/status.svg)](https://doi.org/10.21105/joss.00719)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=
